### PR TITLE
Adds view sensor

### PR DESCRIPTION
**Pull request by SmArtKar**
## About The Pull Request

Adds view sensor component to circuits which allows to get all visible(no ghost detectors, sorry) movable atoms in a 5 tile range. It only works when your circuit's shell is placed on the ground or is held in hands.
Watermelon was fine with the idea.

## Why It's Good For The Game
 
Allows people to construct elaborate machines, dunno.

## Changelog

:cl:
expansion: Added new view sensor component.
/:cl:

### CODE REVIEW

File: code/modules/wiremod/components/sensors/view_sensor.dm [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#define VIEW_SENSOR_COOLDOWN 0.5 SECONDS
+
+/obj/item/circuit_component/view_sensor
+	display_name = "View Sensor"
+	desc = "Outputs a list with all movable objects in it's view. Requires a shell."
+	category = "Sensor"
+
+	circuit_flags = CIRCUIT_FLAG_INPUT_SIGNAL|CIRCUIT_FLAG_OUTPUT_SIGNAL
+
+	power_usage_per_input = 5 //Normal components have 1
```
Comment (Watermelon914) [inline code/modules/wiremod/components/sensors/view_sensor.dm:17]: beef up power usage some more, ideally people shouldn't be spamming this component

File: code/modules/wiremod/components/sensors/view_sensor.dm [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+				result.set_output(null)
+				return
+		else
+			result.set_output(null)
+			return
+
+	var/object_list = list()
+
+	for(var/atom/movable/target in view(5, get_turf(parent.shell)))
+		if(target.invisibility > SEE_INVISIBLE_LIVING)
```
Comment (Watermelon914) [inline code/modules/wiremod/components/sensors/view_sensor.dm:47]: make this a variable, it'd be funny if admins could vv it to let it see ghosts

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/modules/research/techweb/all_nodes.dm [hunk 3]
```diff
@@ -289,6 +289,7 @@
 		"comp_trigonometry",
 		"comp_typecast",
 		"comp_typecheck",
+		"comp_view_sensor",
```
[the remark this answers is not in the document]
Comment (SmArtKar) [reply code/modules/research/techweb/all_nodes.dm:292]: It is? It's T, U, V

File: code/modules/wiremod/components/sensors/view_sensor.dm [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+	circuit_flags = CIRCUIT_FLAG_INPUT_SIGNAL|CIRCUIT_FLAG_OUTPUT_SIGNAL
+
+	power_usage_per_input = 10 //Normal components have 1
+
+	/// The result from the output
+	var/datum/port/output/result
+	var/datum/port/output/cooldown
+
+	var/use_cooldown = 0
```
Comment (Fikou) [inline code/modules/wiremod/components/sensors/view_sensor.dm:23]: this should use cooldown defines

File: code/modules/wiremod/components/sensors/view_sensor.dm [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	COOLDOWN_DECLARE(use_cooldown)
+
+	var/see_invisible = SEE_INVISIBLE_LIVING
+
+/obj/item/circuit_component/view_sensor/populate_ports()
+	result = add_output_port("Result", PORT_TYPE_LIST(PORT_TYPE_ATOM))
+	cooldown = add_output_port("Scan On Cooldown", PORT_TYPE_SIGNAL)
+
+/obj/item/circuit_component/view_sensor/input_received(datum/port/input/port)
+	if(!COOLDOWN_FINISHED(src, use_cooldown))
```
Comment (Watermelon914) [inline code/modules/wiremod/components/sensors/view_sensor.dm:32]: this cooldown needs to be circuit based, or else you can just put multiple of the same components in the same circuit and bypass cooldown

**Pull request merged**